Kernel/SanCov.cpp

```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright (c) 2021, Patrick Meyer <git@the-space.agency>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <Kernel/Devices/KCOVDevice.h>
+#include <Kernel/Process.h>
+#include <Kernel/Thread.h>
+
+extern bool g_in_early_boot;
+
+extern "C" {
+
+void __sanitizer_cov_trace_pc(void);
+void __sanitizer_cov_trace_pc(void)
+{
+    if (g_in_early_boot) [[unlikely]]
+        return;
+
+    if (Processor::current().in_irq()) [[unlikely]] {
+        // Do not trace in interrupts.
+        return;
+    }
+
+    auto thread = Thread::current();
+    auto tid = thread->tid();
+    auto maybe_kcov_instance = KCOVDevice::thread_instance->get(tid);
+    if (!maybe_kcov_instance.has_value()) [[likely]] {
```

Comment:
On x86_64, The kernel crashes here without line 28 and below commented out

```
[#0 SystemServer(6:6)]: mount: successfully mounted /dev/pts
SystemServer(6:6): Spawning mount -a to mount all filesystems.
[#0 mount(8:8)]: Unrecoverable page fault, write to address V0x0000100000000070
[mount(8:8)]: CRASH: CPU #0 Page Fault in ring 0
[#0 mount(8:8)]: Exception code: 0002 (isr: 0000)
[#0 mount(8:8)]:     pc=0x0008:0x000000200094733d rflags=0x0000000000010046
[#0 mount(8:8)]:  stack=0x00000020059f27c0
[#0 mount(8:8)]:    rax=0x000000200095002b rbx=0x00000fffffffffc0 rcx=0x0000000000000000 rdx=0x00000020059f2a68
[#0 mount(8:8)]:    rbp=0x00000020059f28b0 rsp=0x00000020059f27c0 rsi=0x0000100000000070 rdi=0x00000020059f2818
[#0 mount(8:8)]:     r8=0x0000002000b38720  r9=0x0000000000000005 r10=0x0000000000000000 r11=0x0000000000000003
[#0 mount(8:8)]:    r12=0x00000020059f29d8 r13=0x00000020059f2968 r14=0x0000100000000080 r15=0x0000100000000078
[#0 mount(8:8)]:    cr0=0x0000000080010013 cr2=0x0000100000000070 cr3=0x0000000004296000 cr4=0x00000000003406e4
[mount(8:8)]: KERNEL PANIC! :^(
[mount(8:8)]: Crash in ring 0
[mount(8:8)]: at ../../Kernel/Arch/x86/common/Interrupts.cpp:228 in void Kernel::handle_crash(const Kernel::RegisterState&, const char*, int, bool)
```
Are we maybe missing some extra "dont sancov me"  CMake annotations for x86_64?

Reply:
Yep, seems the culprit is Arch/x86/x86_64/PRocessor.cpp.

Adding this fixes the boot and lets the coverage work out

```
    if ("${SERENITY_ARCH}" STREQUAL "x86_64")
        set_source_files_properties(../Kernel/Arch/x86/x86_64/Processor.cpp PROPERTIES COMPILE_FLAGS "-fno-sanitize-coverage=trace-pc")
    endif()
```
Or I guess... this would work too?

```
set_source_files_properties(../Kernel/Arch/x86/${KERNEL_ARCH}/Processor.cpp PROPERTIES COMPILE_FLAGS "-fno-sanitize-coverage=trace-pc")
```

Reply:
Thanks for looking into this :) I've added the second version!

It would probably be good to add at least basic tests somewhere for the debug build..